src/llama-sampling.cpp

```diff
@@ -1059,6 +1059,86 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
     };
 }
 
+// xtc
+
+struct llama_sampler_xtc {
+    const float  probability;
+    const float  threshold;
+    const float  threshold_max;
+    const size_t min_keep;
+};
+
+static const char * llama_sampler_xtc_name(const struct llama_sampler * /*smpl*/) {
+    return "xtc";
+}
+
+static void llama_sample_xtc_apply(struct llama_sampler * smpl, llama_token_data_array * cur_p) {
+    const auto * ctx = (llama_sampler_xtc *) smpl->ctx;
+
+    if (ctx->probability <= 0.0f || ctx->threshold <= 0.0f || cur_p->size <= 1 || ctx->min_keep <= 2) {
+        return;
+    }
+
+    std::random_device rd;
+    float chance = (float)(rd()%100)/100;
+    if (chance > ctx->probability) return;
+    // in case it's not sorted/recalculated yet
+    llama_sampler_softmax_impl(cur_p);
+
+    int removed = 0;
+    // going through all candidates from back to front, easier to keep the last of probables
+    for (int i = (cur_p->size - 1); i >= 0; --i) {
```

Comment:
I kept this version since XTC is not meant to be used before truncating samplers. The recommended sequence is "min-p" -> "xtc", so the amount of candidates should not be too long. In practice, I have not noticed it affecting tg speed even with no prior truncation. Plus, this version keeps the code a bit cleaner.

As for sorting, it is the same as in `llama_sampler_softmax_impl`. Your idea is good though, I will look into that.

Reply:
I'm quite sure performance will be visibly affected if there is a high baseline speed and a large vocabulary (some models have 200k tokens in their vocabulary nowadays). If you are seeing 80 t/s normally, then there's no way you can loop over 200k candidates 80 times per second on most hardware, in addition to all the other TG work.

The TGWUI implementation uses PyTorch vector operations for high performance; maybe something equivalent can be done here.

Reply:
> some models have 200k tokens

Normally that amount of token will be pruned first by `min_p`: in my tests it narrowed down to 50-60 tokens at largest. My logics in this implementation was the following: most often we'll get around 3 to 4 candidates, and even if we have only 2 penalizable candidates, that would result in almost the same about of iterations going in one loop as going in two loops (find penalizable ones from the front, then go through them and actually penalize). The more penalizable candidates there are, the more effective it is to go in one loop.

Sorting is inevitable here, so I will try to optimize it.

Reply:
I tried replacing sort with a simple shift, but I'm not sure if it's faster (should be, given `std::sort`'s complexity). Additionally, it presents a problem of duplicated tokens at the back if `min_keep` is above 0.

Is it a better solution?

```C++
    int found = 0;
    int pos = 0;
    // going through all candidates from back to front, easier to keep the last of probables
    for (int i = (cur_p->size - 1); i >= 0; --i) {
        if (cur_p->data[i].p >= ctx->threshold && cur_p->data[i].p <= ctx->threshold_max) {
            ++found;
            if (found > 1) {
                // .logits are used for sorting and calculating .p in llama_sample_softmax_impl
                cur_p->data[i].logit = -999.0f;
                pos = i;
            }
        }
    }

    if (found > 1) {
        int removed = found - 1;
        size_t size_new = cur_p->size - removed;

        for (size_t i = pos; i < size_new - pos; ++i) {
            cur_p->data[i] = cur_p->data[i + removed];
        }

        if (size_new < ctx->min_keep) size_new = ctx->min_keep;
        cur_p->size = size_new;
    }
```

Reply:
I think I've found a solution, it should be faster than the previous one. At thin point, however, it might be worth reworking the first for loop.

Reply:
Looking at:

https://github.com/oobabooga/text-generation-webui/pull/6335

Can't you just use almost the same logic as `min_p`:

```cpp
    // if the cur_p aren't sorted, try the unsorted implementation first
    if (!cur_p->sorted) {
        std::vector<llama_token_data> filtered_tokens;

        float max_logit = -FLT_MAX;
        for (size_t i = 0; i < cur_p->size; ++i) {
            max_logit = std::max(max_logit, cur_p->data[i].logit);
        }
        const float min_logit = max_logit + logf(ctx->p); // min logit for p_i >= p * p_max

        for (size_t i = 0; i < cur_p->size; ++i) {
            if (cur_p->data[i].logit >= min_logit) {
                filtered_tokens.push_back(cur_p->data[i]);
            }
        }

        // if we have enough values the operation was a success
        if (filtered_tokens.size() >= ctx->min_keep) {
            memcpy(cur_p->data, filtered_tokens.data(), filtered_tokens.size()*sizeof(llama_token_data));
            cur_p->size = filtered_tokens.size();
            min_p_applied = true;
        }
    }
 ```

but use `cur_p->data[i].logit <= min_logit` to place into `filtered_tokens`, and then keep track of a single  `best_token` / `best_logit` pair (ie: that is current best that exceeds `min_logit` by the smallest amount). Then at the end prepend `best_logit` before the `memcpy` if found?

Unless I've misunderstood the XTC logic then this seems the simplest way to do it?

Reply:
@jukofyork  While batch-testing in my program I noticed consistently shorter results. Out of worry that it was caused by my current implementation of XTC (it was not, I haven't updated to the latest fixes of the new backend registry) I rewrote it with memcpy similar to what you've proposed:

```C++
    for (int i = 0; i < candidates->size; ++i) {
        if (cur_p->data[i].p < ctx->threshold || cur_p->data[i].p > ctx->threshold_max) {
            if (i > 0 && cur_p->data[i-1].p >= ctx->threshold) cur.emplace_back(cur_p->data[i-1]);
            cur.emplace_back(cur_p->data[i]);
        } else if (pos == -1) pos = i;
    }

    size_t to_remove = cur_p->size - cur.size();

    if (to_remove >= ctx->min && cur.size() >= ctx->min_keep)) {
        memcpy(cur_p->data, cur.data(), cur.size()*sizeof(llama_token_data));
        cur_p->size = cur.size();
    }
```

Is it correct? Am I missing anything?